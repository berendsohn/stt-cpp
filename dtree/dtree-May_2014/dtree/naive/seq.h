// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_SEQ_H_
#define DTREE_NAIVE_SEQ_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndSeq : public Base {
 public:
  EndSeq() : solid_() {}
  EndSeq* solid(int i) const { return solid_[i]; }
  void set_solid(int i, EndSeq* solid_i) { solid_[i] = solid_i; }
  template<typename Forwarder>
  void CopyFrom(const EndSeq& node, Forwarder forward);
 private:
  EndSeq* solid_[2];
  DISALLOW_COPY_AND_ASSIGN(EndSeq);
};

template<typename Group, typename Base>
class WithValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToSeq(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromSeq(Node* node, typename Group::Type delta);

 protected:
  WithValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  template<typename Node>
  static typename Semigroup::Type AggrSeq(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostSeq(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeftmostSeq(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRightmostSeq(Node* node, const Predicate& predicate);
 protected:
  WithAggr() {}
};

template<typename Base>
class WithReverseBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void ReverseSeq(Node* node);
 protected:
  WithReverseBy() {}
};

template<typename Base>
class WithReverse : public Base {
 public:
  template<typename Node>
  static void ReverseSeq(Node* node);
 protected:
  WithReverse() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_SEQ_H_
