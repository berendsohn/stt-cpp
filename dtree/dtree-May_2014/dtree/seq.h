// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_SEQ_H_
#define DTREE_SEQ_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/seq.h"
#else
#include "dtree/selfadjust/seq.h"
#endif
#endif  // DTREE_SEQ_H_
