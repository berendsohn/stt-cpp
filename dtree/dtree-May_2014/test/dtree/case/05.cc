#define DTREE_CONFIG EndSeq<WithReverse<WithReverse<WithValue<FreeGroup, WithAggr<dtree::Count<int>, dtree::Index_<int>, WithAggr<dtree::Count<int>, dtree::Index_<int>, Begin > > > > > >
#include "test/dtree/seq.cc"
