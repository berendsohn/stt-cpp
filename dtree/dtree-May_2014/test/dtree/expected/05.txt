<<<
  12: v0a1,000,940,000
  2: v038,000,618,000
  13: vd3e,000,244,000
  9: ve74,000,848,000
  6: vbd5,000,5e2,000
  3: va3d,000,349,000
  10: v5dc,000,d17,000
  1: v89a,000,7cb,000
  7: vda8,000,813,000
  4: vea5,000,b78,000
  15: v6ef,000,362,000
  8: v48b,000,d62,000
  11: v1d1,000,b52,000
  14: va48,000,61c,000
  5: vb3e,000,b88,000
>>>
<<<
      /14: a11 a11 va48,d3e,61c,244 f0 f0
              /4: a1 a1 vea5,6ef,b78,362 f0 f1
            /15: a2 a2 v6ef,da8,362,813 f1 f0
          /7: a8 a8 vda8,48b,813,d62 f0 f1
            \9: a5 a5 ve74,da8,848,813 f0 f1
                  /10: a1 a1 v5dc,89a,d17,7cb f1 f1
                /1: a2 a2 v89a,a3d,7cb,349 f1 f0
              \3: a4 a4 va3d,e74,349,848 f0 f1
                \6: a1 a1 vbd5,a3d,5e2,349 f1 f1
        \8: a10 a10 v48b,a48,d62,61c f0 f1
          \11: a1 a1 v1d1,48b,b52,d62 f1 f0
    /13: a14 a14 vd3e,b3e,244,b88 f0 f1
        /2: a1 a1 v038,0a1,618,940 f0 f0
      \12: a2 a2 v0a1,d3e,940,244 f0 f0
  @5: a12 a12 vb3e,000,b88,000 f0 f0
>>>
